### PR TITLE
Sanity checks for function evaluation.

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs [hunk 1]
```diff
@@ -454,6 +464,13 @@ matchVariableFunction
                     ]
 matchVariableFunction _ _ _ _ = nothing
 
+data ArbitrarySymbols a = ArbitrarySymbols
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Does "arbitrary" here mean "not constructor-like"?

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs [hunk 2]
```diff
@@ -478,14 +495,25 @@ matchNonVarToPattern
...
                 tools substitutionSimplifier first second
         (return . return)
             (fmap secondVariablesSubstitutionToPredicate result)
+    let
+        firstVariables = freePureVariables first
+        resultVariables =
+            foldMap
+                (freeVariablesUnderArbitrarySymbols tools)
+                (OrOfExpandedPattern.extractPatterns finalResult)
+    if null (Set.intersection firstVariables resultVariables)
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Could you add a comment here explaining what the logic is? I'm struggling to understand the big picture here.

Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Thanks, I should have done that before sending the code review. Is it easy enough to understand now?

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs (hunk 1 above)
Comment (virgil-serbanuta) [reply src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: Well... Yes. Or almost, it means not constructor-modulo-like. I renamed it.

**Comment by traiansf** [comment]
I'm OK with it in general, and when @ttuegel approves it you can merge it, but I was thinking this might deserve a design decision doc, just in case it stays in the codebase long enough.

File: src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs (hunk 2 above)
Comment (ttuegel) [reply src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs]: :+1

**Pull request merged**